plugins/rts54xx/fu-rts54xx-device.c

```diff
@@ -0,0 +1,411 @@
+{
+	const guint8 cmd[] = { 0x40, 0x06, val, 0x00, 0x00, 0x00, 0x00, 0x00 };
+	guint8 buf[FU_RTS54XX_DEVICE_REPORT_LENGTH] = { 0 };
+	memcpy (buf + FU_RTS54XX_DEVICE_REPORT_CMD_OFFSET, cmd, sizeof(cmd));
+	if (!fu_rts54xx_device_set_report (self, buf, sizeof(buf), error)) {
+		g_prefix_error (error, "failed to set clock-mode=%i: ", val);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_rts54xx_device_reset_to_flash (FuRts54xxDevice *self, GError **error)
+{
+	const guint8 cmd[] = { 0x40, 0xe9, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
+	guint8 buf[FU_RTS54XX_DEVICE_REPORT_LENGTH] = { 0 };
+	memcpy (buf + FU_RTS54XX_DEVICE_REPORT_CMD_OFFSET, cmd, sizeof(cmd));
+	if (!fu_rts54xx_device_set_report (self, buf, sizeof(buf), error)) {
+		g_prefix_error (error, "failed to soft reset: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_rts54xx_device_write_flash (FuRts54xxDevice *self,
+				   guint32 addr,
+				   const guint8 *data,
+				   guint8 data_sz,
+				   GError **error)
+{
+	guint8 cmd[] = { 0x40, 0xc8, 0xff, 0xff, 0xff, 0xff, data_sz, 0x00 };
+	guint8 buf[FU_RTS54XX_DEVICE_REPORT_LENGTH] = { 0 };
+
+	g_return_val_if_fail (data_sz <= 128, FALSE);
+	g_return_val_if_fail (data != NULL, FALSE);
+	g_return_val_if_fail (data_sz != 0, FALSE);
+
+	fu_common_write_uint32 (cmd + 2, addr, G_BIG_ENDIAN);
+	memcpy (buf + FU_RTS54XX_DEVICE_REPORT_CMD_OFFSET, cmd, sizeof(cmd));
+	memcpy (buf + FU_RTS54XX_DEVICE_REPORT_DATA_OFFSET, data, data_sz);
+	if (!fu_rts54xx_device_set_report (self, buf, sizeof(buf), error)) {
+		g_prefix_error (error, "failed to write flash @%08x: ", (guint) addr);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_rts54xx_device_verify_update_fw (FuRts54xxDevice *self, GError **error)
+{
+	const guint8 cmd[] = { 0x40, 0xd9, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00};
+	guint8 outbuf[1];
+
+	/* read command */
+	if (!fu_rts54xx_device_send_in (self,
+					cmd, sizeof(cmd),
+					NULL, 0, /* data */
+					outbuf, sizeof(outbuf),
+					4, error))
+		return FALSE;
+
+	/* check device status */
+	if (outbuf[0] != 0x00) { /* FIXME, what value? */
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_WRITE,
+				     "firmware flash failed");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_rts54xx_device_erase_spare_bank (FuRts54xxDevice *self, GError **error)
+{
+	const guint8 cmd[] = { 0x40, 0xe8, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00};
+	guint8 buf[FU_RTS54XX_DEVICE_REPORT_LENGTH] = { 0 };
+	memcpy (buf + FU_RTS54XX_DEVICE_REPORT_CMD_OFFSET, cmd, sizeof(cmd));
+	if (!fu_rts54xx_device_set_report (self, buf, sizeof(buf), error)) {
+		g_prefix_error (error, "failed to erase spare bank: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_rts54xx_device_ensure_status (FuRts54xxDevice *self, GError **error)
+{
+	const guint8 cmd[] = { 0xc0, 0x09, 0x00, 0x00, 0x00, 0x00, 12, 0x00 };
```

Comment:
I think it would be good to document in this plugin the form of this structure.  (for example looking at this plugin first byte `0xc0` looks like it's only for reading and first byte `0x40` looks like it's used for commands or writing).

Reply:
Fixed, thanks. @superm1 I'll add some more #defines too.